rclpy/src/rclpy/qos_events.cpp

```diff
@@ -0,0 +1,351 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcl/error_handling.h>
+#include <rcl/event.h>
+#include <rcl/types.h>
+#include <rcpputils/scope_exit.hpp>
+#include <rmw/incompatible_qos_events_statuses.h>
+
+#include <memory>
+#include <utility>
+
+#include "rclpy_common/common.h"
+#include "rclpy_common/handle.h"
+#include "rclpy_common/exceptions.hpp"
+
+#include "qos_events.hpp"
+
+namespace rclpy
+{
+
+namespace
+{
+
+#define PYMODULE_NAME "rclpy.qos_event"
+
+typedef union qos_event_callback_data {
+  // Subscription events
+  rmw_requested_deadline_missed_status_t requested_deadline_missed;
+  rmw_liveliness_changed_status_t liveliness_changed;
+  rmw_message_lost_status_t message_lost;
+  rmw_requested_qos_incompatible_event_status_t requested_incompatible_qos;
+  // Publisher events
+  rmw_offered_deadline_missed_status_t offered_deadline_missed;
+  rmw_liveliness_lost_status_t liveliness_lost;
+  rmw_offered_qos_incompatible_event_status_t offered_incompatible_qos;
+} qos_event_callback_data_t;
+
+typedef py::object qos_event_data_filler_function (const qos_event_callback_data_t *);
+
+template<typename T>
+using unique_cstruct_ptr = std::unique_ptr<T, void (*)(T *)>;
+
+unique_cstruct_ptr<rcl_event_t>
+create_zero_initialized_event()
+{
+  unique_cstruct_ptr<rcl_event_t> event(
+    static_cast<rcl_event_t *>(PyMem_Malloc(sizeof(rcl_event_t))),
+    [](rcl_event_t * event) {
+      if (!event) {
+        PyErr_Clear();
+        int stack_level = 1;
+        PyErr_WarnFormat(
+          PyExc_RuntimeWarning, stack_level,
+          "no rcl_event_t to delete");
+      }
+      rcl_ret_t ret = rcl_event_fini(event);
+      if (RCL_RET_OK != ret) {
+        int stack_level = 1;
+        PyErr_WarnFormat(
+          PyExc_RuntimeWarning, stack_level,
+          "failed to fini event: %s",
+          rcl_get_error_string().str);
+        rcl_reset_error();
+      }
+      PyMem_Free(event);
+    });
+  if (!event) {
+    throw std::bad_alloc();
+  }
+  *event = rcl_get_zero_initialized_event();
+  return event;
+}
+
+rcl_publisher_event_type_t
+rcl_publisher_event_type_from(py::object pyevent_type)
+{
+  py::module qos_events = py::module::import(PYMODULE_NAME);
+  py::object pyevent_type_enum = qos_events.attr("QoSPublisherEventType");
```

Comment:
Ah looks like the [enum definition in Python duplicates the constants](https://github.com/ros2/rclpy/blob/41f9c2fea1fc82e7383c53689fdb2c2f74e03d47/rclpy/rclpy/qos_event.py#L31-L40). I'd recommend defining the enums in Pybind11 to avoid both that and avoid having to import the type here.


See `ClockType` as an example

https://github.com/ros2/rclpy/blob/4d76c3a522d1ffdffcb2a5e7d925781f01fbfa07/rclpy/src/rclpy/_rclpy_pybind11.cpp#L38-L42
https://github.com/ros2/rclpy/blob/4d76c3a522d1ffdffcb2a5e7d925781f01fbfa07/rclpy/rclpy/clock.py#L21

Reply:
This is also nice to have, but we could open an issue for it and move on for now.

Reply:
Done in https://github.com/ros2/rclpy/pull/723/commits/2795b162bed6b8bb5b5d4dbbf6e8a435258702c4.